lib/engine/step/g_1860/tracker.rb

```diff
@@ -0,0 +1,279 @@
+
+          return unless terrain.any?
+
+          @game.all_companies_with_ability(:tile_income) do |company, ability|
+            if terrain.include?(ability.terrain) && (!ability.owner_only || company.owner == entity)
+              # If multiple borders are connected bonus counts each individually
+              income = ability.income * terrain.find_all { |t| t == ability.terrain }.size
+              @game.bank.spend(income, company.owner)
+              @log << "#{company.owner.name} earns #{@game.format_currency(income)}"\
+                " for the #{ability.terrain} tile built by #{company.name}"
+            end
+          end
+        end
+
+        def border_cost(tile, entity)
+          hex = tile.hex
+          types = []
+
+          total_cost = tile.borders.dup.sum do |border|
+            next 0 unless (cost = border.cost)
+
+            edge = border.edge
+            neighbor = hex.neighbors[edge]
+            next 0 if !hex.targeting?(neighbor) || !neighbor.targeting?(hex)
+
+            types << border.type
+            tile.borders.delete(border)
+            neighbor.tile.borders.map! { |nb| nb.edge == hex.invert(edge) ? nil : nb }.compact!
+
+            ability = entity.all_abilities.find do |a|
+              (a.type == :tile_discount) && (border.type == a.terrain)
+            end
+            discount = ability&.discount || 0
+
+            if discount.positive?
+              @log << "#{entity.name} receives a discount of "\
+                "#{@game.format_currency(discount)} from "\
+                "#{ability.owner.name}"
+            end
+
+            cost - discount
+          end
+          [total_cost, types]
+        end
+
+        def check_track_restrictions!(entity, old_tile, new_tile, old_revenues)
+          return if @game.loading || !entity.operator?
+
+          changed_city = false
+          if old_tile.color != :white
+            # add requirement that paths/nodes be reachable with train
+            @game.game_error('Tile must be reachable with train') unless reachable_hex?(entity, new_tile.hex)
+            new_revenues = new_tile.nodes.select { |n| reachable_node?(entity, n) }.map(&:max_revenue).sort
+            changed_city = old_revenues != new_revenues
+          end
+
+          old_paths = old_tile.paths
+          used_new_track = old_paths.empty?
+
+          new_tile.paths.each do |np|
+            next unless @game.graph.connected_paths(entity)[np]
+            next if old_tile.color != :white && !reachable_path?(entity, np)
+
+            op = old_paths.find { |path| np <= path }
+            used_new_track = true unless op
+
+            next unless old_tile.color == :white
+
+            old_revenues = op&.nodes && op.nodes.map(&:max_revenue).sort
+            new_revenues = np&.nodes && np.nodes.map(&:max_revenue).sort
+            changed_city = true unless old_revenues == new_revenues
+          end
+
+          case @game.class::TRACK_RESTRICTION
```

Comment:
is this all copy and pasted?

Reply:
No. See above.

Reply:
above where?

Reply:
The code snippet above this one. Tracker was copied over because I had to change some, but not all, of it's methods. I've since changed it so that the 1860-specific Track includes the generic one and overloads the changed methods.

Reply:
Specifically, I had to modify check Tracker::check_track_restrictions! to compare reachable revenues before and after the tile lay. Unfortunately this meant having to make a very small modification to lay_tile_action.